src/Database/LSMTree/Internal/Chunk.hs

```diff
@@ -0,0 +1,98 @@
+module Database.LSMTree.Internal.Chunk
+(
+    -- * Chunks
+    Chunk (Chunk),
+
+    -- * Balers
+    Baler,
+    createBaler,
+    feedBaler,
+    readBalerRemnant
+)
+where
+
+import           Prelude hiding (concat, length, splitAt)
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.STRef.Strict (STRef, newSTRef, readSTRef, writeSTRef)
+import           Data.Vector.Primitive (Vector, concat, length, slice, splitAt)
+import           Data.Word (Word8)
+
+-- * Chunks
+
+-- | A chunk of bytes, typically output during incremental index serialisation.
+newtype Chunk = Chunk (Vector Word8)
+
+-- * Balers
+
+-- Mutable state of a baler.
+data BalerState = BalerState
+                      ![Vector Word8] -- Reverse list of queued blocks
+                      !Int            -- Total size of queued blocks
+
+{-|
+    An object that receives blocks of bytes and repackages them into chunks of
+    preferably a given maximum size.
+-}
+data Baler s = Baler
+                   !Int                  -- Maximum chunk size
+                   !(STRef s BalerState) -- Reference to the mutable state
+
+-- | Creates a new baler.
+createBaler :: Int            -- ^ Maximum chunk size
+            -> ST s (Baler s) -- ^ Creation of the baler
+createBaler maxChunkSize = Baler maxChunkSize <$> newSTRef (BalerState [] 0)
+
+{-|
+    Feeds a baler a block of bytes.
+
+    Bytes received by a bailer are generally queued for later output, but if
+    feeding new bytes makes the queued content exceed the maximum chunk size
+    then as many chunks of maximum size as possible are output.
+-}
+feedBaler :: Vector Word8 -> Baler s -> ST s [Chunk]
+feedBaler block (Baler maxChunkSize stateRef) = do
+    BalerState remnantBlocksRev remnantSize <- readSTRef stateRef
+    let
+
+        totalSize :: Int
+        totalSize = remnantSize + length block
+
+    if totalSize < maxChunkSize
+        then do
+                 writeSTRef stateRef
+                            (BalerState (block : remnantBlocksRev) totalSize)
```

Comment:
```suggestion
                 writeSTRef stateRef $!
                            (BalerState (block : remnantBlocksRev) totalSize)
```

Reply:
Wouldn’t this use of `($!)` be superfluous? The expression of what to put into the mutable variable is an application of a data constructor and therefore in head normal form. No thunk should be created; the value should be in evaluated form from the outset.

Reply:
Hmm, on second thought I’m not so sure anymore. The point is that `BalerState` has strict fields and therefore is not an ordinary data constructor. In fact, it should effectively be defined internally as follows, where `BalerState'` denotes the actual, underlying data constructor:
```
BalerState remnantBlocksRev remnantSize = remnantBlocksRev `seq` remnantSize `seq` BalerState' remnantBlocksRev remnantSize
```
Ironically, this would make the above `writeSTRef` application less strict in a sense, as its second argument wouldn’t be a bare data constructor application anymore. Is this how it is?

Reply:
I know that `writeSTRef` is not strict. My point was that a mere application of a data constructor would be in head normal form from the start, so that there would be nothing to evaluate, although my later considerations above suggest that this is likely to be different for data constructors with strict fields. Because of the latter, adding `$!` seems to make sense indeed.

Reply:
The `writeSTRef` strictness issue is now fixed in 43c4f491b1d6b9460294cf3aed6c5c5b85c55f83.